### PR TITLE
Autodetect and copy installed heroes on windows

### LINKED ISSUE #3437: VCMI first install: read default H3 install path from registry on Windows
As in title - when you import H3 data via welcome screen on Windows, default location on opening file picker should show H3 installation folder read from Windows registry.

Hota does this and it makes installing easier.

### PR DESCRIPTION
**Pull request by Laserlicht**
fixes  #3437

### CODE REVIEW

**Comment by dydzio0614** [comment]
hopefully somebody can test this manually (no windows here)

**Comment by Laserlicht** [comment]
> hopefully somebody can test this manually (no windows here)

Tested the first commit. Will retest after CI is done.

But would be great if someone other than me also could test.

**Comment by Laserlicht** [comment]
Tested following on my Win7 32bit VM:
- Installed CD version
- Installed GOG version
- Without installed H3 (data in custom folder)

Works!

**Review by kambala-decapitator — APPROVED** [review]
use `{}` or `QString{}` instead of empty string literals

**Comment by IvanSavenko** [comment]
Perhaps show confirmation query instead of silent copy?

Not against silent copy, but not sure if this is a good idea.

**Comment by Laserlicht** [comment]
Added Question-Box

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 1]
```diff
@@ -254,9 +266,25 @@ void FirstLaunchView::forceHeroesLanguage(const QString & language)
...
+	return QString{};
+}
+
+void FirstLaunchView::copyHeroesData(const QString & path)
 {
-	QDir sourceRoot = QFileDialog::getExistingDirectory(this, "", "", QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks);
+	QDir sourceRoot = QDir(path);
+	
+	if(path.isEmpty())
+		sourceRoot.setPath(QFileDialog::getExistingDirectory(this, "", "", QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks));
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:287]:
```suggestion
		sourceRoot.setPath(QFileDialog::getExistingDirectory(this, {}, {}, QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks));
```

File: launcher/firstLaunch/firstlaunch_moc.h [hunk 2]
```diff
@@ -51,7 +51,8 @@ class FirstLaunchView : public QWidget
 	void heroesLanguageUpdate();
 	void forceHeroesLanguage(const QString & language);
 
-	void copyHeroesData();
+	QString getHeroesInstallDir();
+	void copyHeroesData(const QString & path = "");
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.h:55]:
```suggestion
	void copyHeroesData(const QString & path = {});
```

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 3]
```diff
@@ -254,9 +266,25 @@ void FirstLaunchView::forceHeroesLanguage(const QString & language)
 	node->String() = language.toStdString();
 }
 
-void FirstLaunchView::copyHeroesData()
+QString FirstLaunchView::getHeroesInstallDir()
+{
+#ifdef VCMI_WINDOWS
+	QString gogPath = QSettings("HKEY_LOCAL_MACHINE\\SOFTWARE\\GOG.com\\Games\\1207658787", QSettings::NativeFormat).value("path").toString();
+	if(!gogPath.isEmpty())
+		return gogPath;
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:274]:
```suggestion
		return gogPath;

```

File: launcher/firstLaunch/firstlaunch_moc.cpp [hunk 4]
```diff
@@ -134,7 +134,17 @@ void FirstLaunchView::activateTabHeroesData()
...
 	}
-	heroesDataUpdate();
+	if(heroesDataUpdate())
+		return;
+
+	QString installPath = getHeroesInstallDir();
+	if(!installPath.isEmpty())
+	{
+		QMessageBox::StandardButton reply;
+		reply = QMessageBox::question(this, tr("Heroes III installation found!"), tr("Copy data to VCMI folder?"), QMessageBox::Yes|QMessageBox::No);
```
Comment (kambala-decapitator) [inline launcher/firstLaunch/firstlaunch_moc.cpp:144]:
```suggestion
		auto reply = QMessageBox::question(this, tr("Heroes III installation found!"), tr("Copy data to VCMI folder?"), QMessageBox::Yes | QMessageBox::No);
```

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

**Comment by Laserlicht** [comment]
@kambala-decapitator Thanks. :)

**Review by IvanSavenko — APPROVED** [review]

**Comment by IvanSavenko** [comment]
@kambala-decapitator is everything OK now?

**Comment by kambala-decapitator** [comment]
> @kambala-decapitator is everything OK now?

Yep, thanks for reminding

**Comment by IvanSavenko** [comment]
Not related to this PR strictly speaking - this is how our setup works at the moment. We ask player to point to location of H3 files and copy them.
The only change from this PR is that on Windows vcmi will now automatically find this location.

As for your question - partially historical reasons, partially - this approach is easier:
1) This way we don't have to ask player whether to copy or use data as it. 
2) While on Windows H3 files will probably stay, on other platforms they likely won't - since player likely pointed to files that he have extracted into some temporary directory.
3) Support for using data in-place needs work. Likely easy to do, but that is still time.
4) Frankly, this is 2024, not 2000. Most of players have large enough storage to not care about 1 extra Gb. And if they do - they can create symlinks (or whatever they are called on Windows).

**Comment by kambala-decapitator** [comment]
> Why copying files?
> 
> Why not just use the resources from there, where is detected?

We have #874 for that

**Comment by psyhlo** [comment]
thanks and sorry for the spam

**Pull request merged**